### ISSUE TITLE
Random disconnects and stops reconnecting

### ISSUE THREAD

**Issue by battuashwik** [primus/primus#491]
Sometimes the client randomly disconnects and stops reconnecting.

`client = new Socket(socketUrl, {
      ping: 15000,
      pong: 4000,
      reconnect: {
        min: 4000,
        max: 4000,
        retries: 100000000000000
      }
  })
`
These are my connection params. Also on the server side i whenever i send disconnect (if authentication fails) i send 
`spark.end(undefined, { reconnect: true });`

It does work most of the time but suddenly it breaks, like once in 3-4 days or sometimes takes even longer. This is really painful as i need to manually restart client whenever this happens.

**Comment by lpinca** [comment]
This issue report is not really helpful, please give us more details:

- What version of Node.js are you using?
- What version of Primus are you using?
- What transformer are you using?
- Are you using the Node.js or browser client?
- Did you try to use a different transformer?

**Comment by battuashwik** [comment]
Node Version: 6.2.1
Primus Version: 6.0.1
Transformer: Engine.io
Nodejs client
No, I dint try with different transformer.

Is it known to have issues with engine.io?

**Comment by lpinca** [comment]
It could be, can you try with `sockjs` and see if the issue still happens?
Thanks.

**Comment by lpinca** [comment]
Also, if you are using only the Node.js client it might make more sense to use a WebSocket only transformer.

**Comment by battuashwik** [comment]
My use case is only to communicate from node client only, so do u suggest me to use uws transformer?

**Comment by lpinca** [comment]
Yes, `faye`, `uws` or `ws` unless there is a firewall or something that block the WebSocket protocol.

**Comment by lpinca** [comment]
@battuashwik any update?

**Comment by battuashwik** [comment]
Made the change but need more testing, will update by end of this week. The problem with this issue is that it cant be detected while developing.

**Comment by dotob** [comment]
hi, i have kind of the same problem. changed from ws to uws but still exists. did try different version of node/primus. seems always the same. could be a problem of mine. but i wanted to know if you can point me to some things i can do to detect whats the problem. any way to get some info from primus or transformer what happened? any help is aprreciated.

**Comment by lpinca** [comment]
@dotob just to be sure we are on the same line, are you also using the Primus client in Node.js?

**Comment by dotob** [comment]
sorry. forgot to mention. node-node. yes. 

...

**Comment by lpinca** [comment]
On the server you can enable debug logs using the `DEBUG` env variable for example:

```
DEBUG=primus:* node server.js
```

On the client I suggest to add listeners for the recovery [events](https://github.com/unshiftio/recovery#events) which are forwarded to client instance and log what happens.

```js
primus.on('reconnect scheduled', (state) => { ... });
primus.on('reconnect failed', (state) => { ... });
primus.on('reconnect', (state) => { ... });
```

**Comment by dotob** [comment]
thanks. will try it.

**Comment by lpinca** [comment]
Any update here?

**Comment by utan** [comment]
I do have disconnects and won't reconnect client.. Mobile devices though..
Using current primus and socketjs..

At server side I have timeout false.. and its when it disconnect and won't reconnect.. if default there is timeout every moment and reconnects fine..

The timeout emited from server didn't fix disconnections in Mobile devices..

**Comment by lpinca** [comment]
@utan did you update the options? The server-side `timeout` option is no longer valid in `primus@7`. It has been replaced with `pingInterval`. Same is for the `ping` and `pong` options on the client (replaced by `pingTimeout`).

**Comment by utan** [comment]
@lpinca If I leave by default, then keep alive is sent from server right? if it is then I still got the problem or do I need to activate it somehow?

The heartbeats-and-latency section doesn't say anything about the new options..
What would be the options to use at server side? I am also using fresh copy of client made by primus instance so new settings should be updated..

 regards.

**Comment by lpinca** [comment]
Yes, by default it is enabled.

> If needed to deactivate it how is done, pingInterval to false would do?

Correct, setting it to `false` disables the heartbeats. Rebuild the client or set `pingTimeout` to `false` on the client if you do that.

Anyway ideally you shouldn't disable the heartbeats and the client should reconnect when `pingTimeout` expires.

**Comment by utan** [comment]
I see, I will rebuilt the client instead of setting `pingTimeout` to false in client.. 
also maybe an update to docs

Thanks @lpinca , I will start testing..

**Comment by utan** [comment]
@lpinca, can u tell me when u use a mobile device browser say a phone, doesn't disconnect after ur phone screen goes to sleep or black?
because I tested new primus and still got disconnect even if `pingInterval` set to false.

**Comment by lpinca** [comment]
I did a little test on Safari mobile (iPhone). It disconnects when the browser is not on focus or the phone goes to sleep but reconnection seems to work. When the browser is back on focus a new connection is established.

**Comment by utan** [comment]
I know, thing is my app is a chat system which tells people when they connect and disconnect..

This behavior flows the chat with mobile devices it just awfull..

**Comment by lpinca** [comment]
I guess the connection is closed by the browser if it is in background or the phone goes to sleep. In this case I don't think there is much we can do.

**Comment by utan** [comment]
Yeah, thanks for help though.
Regards.

**Comment by webysther** [comment]
I think maybe this https://github.com/sagemathinc/cocalc-docker/issues/194 is related, I paste lots of data in this issue about the problem.
